internal/gengapic/genrest.go

```diff
@@ -163,6 +163,52 @@ type httpInfo struct {
 	verb, url, body string
 }
 
+func (g *generator) pathParams(m *descriptor.MethodDescriptorProto) map[string]*descriptor.FieldDescriptorProto {
+	pathParams := map[string]*descriptor.FieldDescriptorProto{}
+	info, err := getHTTPInfo(m)
+	if info == nil || err != nil {
+		return pathParams
+	}
+
+	// Match using the curly braces but don't include them in the grouping.
+	re := regexp.MustCompile("{([[:alpha:]]+)}")
+	for _, p := range re.FindAllStringSubmatch(info.url, -1) {
+		// In the returned slice, the zeroth element is the full regex match,
+		// and the subsequent elements are the sub group matches.
+		// See the docs for FindStringSubmatch for further details.
+		param := p[1]
+		field := g.lookupField(m.GetInputType(), param)
+		if field == nil {
+			continue
+		}
+		pathParams[param] = field
+	}
+
+	return pathParams
+}
+
+func (g *generator) queryParams(m *descriptor.MethodDescriptorProto) map[string]*descriptor.FieldDescriptorProto {
```

Comment:
Q: Do you plan on calling `pathParams()` and `queryParams()` separately? Since the latter calls the former, the former runs twice. Maybe it's worth renaming the latter `queryAndPathParams()` and returning both? (Note that you'll need to deal with the `info.body` hack below).

Not a biggie, and this can be deferred.

Reply:
I think I like keeping the functions simple and single purpose. If it looks like generator speed is negatively impacted, I'm more inclined to memoize than to combine functionality.

Reply:
Memoizing is a good idea!